src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -3530,75 +3538,88 @@ ${body}
+      size: renderTarget.height * bytesPerRow,
+      usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+    });
 
-  const renderPass = encoder.beginRenderPass({
-    colorAttachments: [
-      {
-        view: renderTarget.createView(),
-        loadOp: 'clear',
-        storeOp: 'store',
-      },
-    ],
-  });
+    const encoder = t.device.createCommandEncoder();
 
-  renderPass.setPipeline(pipeline);
-  renderPass.setBindGroup(0, bindGroup);
-  renderPass.draw(4);
-  renderPass.end();
-  encoder.copyTextureToBuffer(
-    { texture: renderTarget },
-    { buffer: resultBuffer, bytesPerRow },
-    { width: renderTarget.width, height: renderTarget.height }
-  );
-  t.device.queue.submit([encoder.finish()]);
+    const renderPass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          view: renderTarget.createView(),
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+    });
 
-  await resultBuffer.mapAsync(GPUMapMode.READ);
+    renderPass.setPipeline(pipeline!);
+    renderPass.setBindGroup(0, bindGroup);
+    renderPass.draw(4);
+    renderPass.end();
+    encoder.copyTextureToBuffer(
+      { texture: renderTarget },
+      { buffer: resultBuffer, bytesPerRow },
+      { width: renderTarget.width, height: renderTarget.height }
+    );
+    t.device.queue.submit([encoder.finish()]);
 
-  const view = TexelView.fromTextureDataByReference(
-    renderTarget.format as EncodableTextureFormat,
-    new Uint8Array(resultBuffer.getMappedRange()),
-    {
-      bytesPerRow,
-      rowsPerImage: renderTarget.height,
-      subrectOrigin: [0, 0, 0],
-      subrectSize: [renderTarget.width, renderTarget.height],
-    }
-  );
+    await resultBuffer.mapAsync(GPUMapMode.READ);
 
-  let outIdx = 0;
-  const out = new Array<PerTexelComponent<number>>(calls.length);
-  for (const bin of binned) {
-    for (const callIdx of bin) {
-      const x = outIdx % rtWidth;
-      const y = Math.floor(outIdx / rtWidth);
-      out[callIdx] = view.color({ x, y, z: 0 });
-      outIdx++;
+    const view = TexelView.fromTextureDataByReference(
+      renderTarget.format as EncodableTextureFormat,
+      new Uint8Array(resultBuffer.getMappedRange()),
+      {
+        bytesPerRow,
+        rowsPerImage: renderTarget.height,
+        subrectOrigin: [0, 0, 0],
+        subrectSize: [renderTarget.width, renderTarget.height],
+      }
+    );
+
+    let outIdx = 0;
+    const out = new Array<PerTexelComponent<number>>(calls.length);
+    for (const bin of binned) {
+      for (const callIdx of bin) {
+        const x = outIdx % rtWidth;
+        const y = Math.floor(outIdx / rtWidth);
+        out[callIdx] = view.color({ x, y, z: 0 });
+        outIdx++;
+      }
     }
-  }
 
-  renderTarget.destroy();
-  resultBuffer.destroy();
+    resultBuffer.destroy();
 
-  return out;
+    return out;
+  };
+
+  const results = await run(gpuTexture);
+
+  return {
+    run,
+    results,
+    destroy() {
```

Comment:
Q: is this destroy called anywhere now or is meant to be called at a future commit?